mtags/src/main/scala/scala/meta/internal/pc/CompilerAccess.scala

```diff
@@ -1,90 +1,111 @@
+import scala.concurrent.ExecutionContextExecutor
 
 /**
- * Manages the lifecycle of the compiler.
+ * Manages the lifecycle and multi-threaded access to the presentation compiler.
  *
- * - automatically restarts the compiler on expected crashes caused by macroparadise.
- * - handles cancellation via `Thread.interrupt()` to stop the compiler during typechecking.
+ * - automatically restarts the compiler on miscellaneous crashes.
+ * - handles cancellation via `Thread.interrupt()` to stop the compiler during typechecking,
+ *   for functions that support cancellation.
  */
 class CompilerAccess(
+    config: PresentationCompilerConfig,
     sh: Option[ScheduledExecutorService],
     newCompiler: () => MetalsGlobal
-)(implicit ec: ExecutionContext) {
+)(implicit ec: ExecutionContextExecutor) {
   private val logger: Logger = Logger.getLogger(classOf[CompilerAccess].getName)
+
+  private val jobs = CompilerJobQueue()
   private def isEmpty: Boolean = _compiler == null
   private def isDefined: Boolean = !isEmpty
   def reporter: StoreReporter =
     if (isEmpty) new StoreReporter()
     else _compiler.reporter.asInstanceOf[StoreReporter]
-  def shutdown(): Unit = lock.synchronized {
+
+  def shutdown(): Unit = {
+    restart()
+    jobs.shutdown()
+  }
+
+  def restart(): Unit = {
     val compiler = _compiler
     if (compiler != null) {
       compiler.askShutdown()
       _compiler = null
       sh.foreach { scheduler =>
-        scheduler.schedule(new Runnable {
-          override def run(): Unit = {
-            if (compiler.presentationCompilerThread.isAlive) {
-              compiler.presentationCompilerThread.stop()
-            }
+        scheduler.schedule[Unit](() => {
+          if (compiler.presentationCompilerThread.isAlive) {
+            compiler.presentationCompilerThread.stop()
           }
         }, 2, TimeUnit.SECONDS)
       }
     }
   }
 
   /**
-   * Run the given thunk with unique access to the compiler instance.
-   *
-   * Will not run other requests in parallel.
+   * Asynchronously execute a function on the compiler thread with `Thread.interrupt()` cancellation.
    */
-  def withCancelableCompiler[T](
+  def withInterruptableCompiler[T](
       default: T,
       token: CancelToken
-  )(thunk: MetalsGlobal => T): T = lock.synchronized {
-    val thread = Thread.currentThread()
-    Thread.interrupted() // clear interrupt flag
+  )(thunk: MetalsGlobal => T): CompletableFuture[T] = {
     val isFinished = new AtomicBoolean(false)
-    token
-      .onCancel()
-      .whenComplete(new BiConsumer[java.lang.Boolean, Throwable] {
-        override def accept(isCancelled: lang.Boolean, u: Throwable): Unit = {
-          if (isCancelled && isFinished
-              .compareAndSet(false, true) && isDefined) {
+    var queueThread = Option.empty[Thread]
+    val result = onCompilerJobQueue(
+      () => {
+        queueThread = Some(Thread.currentThread())
+        Thread.interrupted() // clear interrupt thread
```

Comment:
shouldn't those be cleared when handling the interruption?

Reply:
They should be, but we have no guarantee that's true. This is only a sanity check to preserve the invariant that each new task doesn't start with the interrupt flag set to true.